[STY] Fix `PTH118` for all files except in `nilearn.datasets`

michellewang opened the pull request:
<!---
This is a suggested pull request template for nilearn.
It's designed to capture information we've found to be useful in reviewing pull requests.

If there is other information that would be helpful to include, please don't hesitate to add it!

Please make sure your pull request also follows the
[contribution guidelines](https://nilearn.github.io/stable/development.html#contribution-guidelines) that
will be enforced during the review process.
-->

<!-- Please indicate after the # which issue you're closing with this PR.
This is helpful for the maintainers AND will magically close the issue when this
pull request is merged!
If the PR closes multiple issues, includes "closes" before each one is listed.
https://help.github.com/articles/closing-issues-using-keywords -->
- Addresses #4327

<!-- Please give a brief overview of what has changed in the PR.
If you're not sure what to write, consider it a note to the maintainers to indicate
what they should be looking for when they review the pull request. -->
Changes proposed in this pull request:

- ~Remove `PTH118` from ruff ignore list~ putting it back because not all errors are handled in this PR
- Find-and-replace `os.path.join` to `Path` from `pathlib` module, then manual checking and optimizing things
- Note: `nilearn.plotting.glass_brain.py` has `from matplotlib.path import Path`, so for this one I did `import pathlib` and used `pathlib.Path`

michellewang commented:
@Remi-Gau if there were functions that returned paths as strings (for example as values in a dictionary), is it right to assume that we would still want to return strings in these functions? Otherwise there would be a lot of tests to change too

Remi-Gau commented:
Yes indeed.

If they are part of the public API let's not change their behavior for now. Maybe later we can think about it.

Otherwise it may be fine to have them return Path but that may affect a lot of other things: maybe check how often it is called before making the change.

In any case if you can keep track of those functions that return strings where path could go, make a list. In this issue or another one.
This way we can go through it later and decide what to do.

michellewang commented:
@Remi-Gau I am thinking of splitting this PR into multiple ones because for some modules there are quite a lot of fixes to do to make all tests pass. So maybe one PR for the easy cases and then one for each module that requires more work. That would make things easier to review. What do you think?

michellewang commented:
Ok so most of the tests were easy to fix but I still think it's worth splitting off at least `nilearn/datasets/func.py` and `nilearn/datasets/tests/test_neurovault.py`. And so far I just tried to make the tests pass but I still need to spend some time and make docstring changes where appropriate

Remi-Gau commented:
> it's worth splitting off at least `nilearn/datasets/func.py` and `nilearn/datasets/tests/test_neurovault.py`

yeah I would put the changes for the `datasets` subpackage in a different PR.

bthirion reviewed (approved):
The PR LGTM.

michellewang commented:
I have removed all changes to `nilearn.datasets` files and cleaned up the rest, so this should be ready to go as long as the tests pass. The `pre-commit` check is failing now because of the remaining `PTH118` errors in `nilearn.datasets`. So I think I will add back `PTH118` to the `pyproject.toml` ignore list for now.

michellewang commented on nilearn/glm/tests/test_first_level.py:
```diff
@@ -1831,6 +1832,16 @@ def test_check_trial_type_warning(tmp_path):
         _check_trial_type([event_file])
 
 
+def test_list_valid_subjects_with_toplevel_files(tmp_path):
+    """Test that only subject directories are returned, not file names."""
+    (tmp_path / "sub-01").mkdir()
+    (tmp_path / "sub-02").mkdir()
+    (tmp_path / "sub-01.html").touch()
+
+    valid_subjects = _list_valid_subjects(tmp_path, None)
+    assert valid_subjects == ["01", "02"]
+
+
```
Adding this test because at one point the changes I had made caused the `visual_testing/reporter_visual_inspection_suite.py` script to fail during documentation building, and there wasn't a test to catch that error

Remi-Gau replied:
nice, thanks

michellewang commented:
Also just wanted to note that I went ahead and fixed some of the other codes too (`open`, `os.path.abspath`, `os.path.dirname`, `glob.glob`) when they made things easier

michellewang commented on nilearn/interfaces/fmriprep/load_confounds_utils.py:
```diff
@@ -230,9 +230,9 @@ def _get_file_name(nii_file):
+        base_dir / crc for crc in confound_file_candidates
     ]
-    found_files = [cr for cr in confound_file_candidates if os.path.isfile(cr)]
+    found_files = [str(cr) for cr in confound_file_candidates if cr.is_file()]
```
Do we want to return a `Path` here since this function is private?

Remi-Gau replied:
I would say so: though once again, depends how many of our things break.

Remi-Gau commented:
> Also just wanted to note that I went ahead and fixed some of the other codes too (`open`, `os.path.abspath`, `os.path.dirname`, `glob.glob`) when they made things easier

yes make sesne

Remi-Gau commented:
> So I think I will add back `PTH118` to the `pyproject.toml` ignore list for now.

Otherwise you can 'file ignore' that error for a certain subpackage or module.

Remi-Gau commented on nilearn/externals/tempita/__init__.py:
Wondering if we should not leave the `externals` untouched and therefore maybe ignore all errors in that part of the code base?

Remi-Gau commented on nilearn/interfaces/fmriprep/load_confounds_utils.py:
```diff
@@ -271,7 +271,7 @@ def get_confounds_file(image_file, flag_full_aroma):
     """Return json data companion file to the confounds tsv file."""
     # Load JSON file
-    return confounds_raw_path.replace("tsv", "json")
+    return str(confounds_raw_path).replace("tsv", "json")
```
get_json is not part of the user facing API so this one as well could return path

Remi-Gau reviewed (approved):
This is good with me.
I think I'd rather merge this as is and maybe tackle some of my minor comments in other small follow up PR if we think they are worth it.
